pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py

```diff
@@ -0,0 +1,643 @@
+"""
+    Base driver for the UHFQA instrument including all common functionality.
+    Application dependent code can be found in the UHFQuantumController and in the
+    UHFQA_qudev modules. 
+"""
```

Comment:
That is a very good point, but I think we should address these issues in a separate ticket.

Reply:
Thank you, Niels, for implementing the feedback. It looks good, except this aspect, where my comment was maybe not clear enough. The UHFQA_core that you have now merged is not functional because of this issue. The Delft-specific child class should work, but an object directly instantiated from the core (or from the qudev child class) won't work. An error will be raised when trying to run an experiment. I have used the following hack in our repository to make it work. Not sure whether there is a cleaner solution.

```
@@ -1102,22 +1102,24 @@ class ZI_base_instrument(Instrument):
             return np.genfromtxt(filename, delimiter=',')
         except OSError as e:
             # if the waveform does not exist yet dont raise exception
             log.warning(e)
             return None
 
     def _length_match_waveforms(self, awg_nr):
         """
         Adjust the length of a codeword waveform such that each individual
         waveform of the pair has the same length
         """
+        if not hasattr(self, '_get_waveform_table'):
+            return
         log.info('Length matching waveforms for dynamic waveform upload.')
         wf_table = self._get_waveform_table(awg_nr)
 
         matching_updated = False
         iter_id = 0
         # We iterate over the waveform table
         while(matching_updated or iter_id == 0):
 
             iter_id += 1
             if iter_id > 10:
                 raise StopIteration
@@ -1196,22 +1198,24 @@ class ZI_base_instrument(Instrument):
         if wf_name not in self._awg_waveforms:
             raise ziConfigurationError(
                 'Trying to mark waveform {} as read-only, but the waveform has not been configured yet!'.format(wf_name))
 
         self._awg_waveforms[wf_name]['readonly'] = True
 
     def _upload_updated_waveforms(self, awg_nr):
         """
         Loop through all configured waveforms and use dynamic waveform uploading
         to update changed waveforms on the instrument as needed.
         """
+        if not hasattr(self, '_get_waveform_table'):
+            return
         # Fixme. the _get_waveform_table should also be implemented for the UFH
         log.info(f"{self.devname}: Using dynamic waveform update for AWG {awg_nr}.")
         wf_table = self._get_waveform_table(awg_nr)
 
         for dio_cw, (wf_name, other_wf_name) in enumerate(wf_table):
             if self._awg_waveforms[wf_name]['dirty'] or self._awg_waveforms[other_wf_name]['dirty']:
                 # Combine the waveforms and upload
                 wf_data = merge_waveforms(self._awg_waveforms[wf_name]['waveform'],
                                           self._awg_waveforms[other_wf_name]['waveform'])
                 # Write the new waveform
                 self.setv(
```

Reply:
See https://github.com/DiCarloLab-Delft/PycQED_py3/pull/657 for a possible solution.